com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaClassFilter.java

```diff
@@ -0,0 +1,306 @@
+        traverse(blackTree.root, 0, wildcardPatterns, new ArrayList<>());
+
+        // Append the hardcode patterns to the result.
+        for (String name : hardcodePatterns) {
+            if (!blackTree.wildcardMatch(name)) {
+                wildcardPatterns.add(name);
+            }
+        }
+
+        return wildcardPatterns.toArray(new String[0]);
+    }
+
+    private static void superimpose(Trie upTree, Trie downTree) {
+        Queue<TrieNode> upQueue = new LinkedList<>();
+        Queue<TrieNode> downQueue = new LinkedList<>();
+        upQueue.offer(upTree.root);
+        downQueue.offer(downTree.root);
+        while (!upQueue.isEmpty()) {
+            TrieNode upNode = upQueue.poll();
+            TrieNode downNode = downQueue.poll();
+            downNode.isGray = true;
+            for (Entry<String, TrieNode> entry : upNode.children.entrySet()) {
+                if (downNode.children.containsKey(entry.getKey())) {
+                    upQueue.offer(entry.getValue());
+                    downQueue.offer(downNode.children.get(entry.getKey()));
+                }
+            }
+        }
+    }
+
+    private static void traverse(TrieNode root, int depth, List<String> result, List<String> parent) {
+        // If the node is gray, that means it also appears in the white list.
+        // We cannot use it to generate a wildcard pattern.
+        if (!root.isGray) {
+            String[] names = new String[depth + 1];
+            for (int i = 0; i < depth - 1; i++) {
+                names[i] = parent.get(i);
+            }
+
+            names[depth - 1] = root.name;
+            names[depth] = "*";
+            result.add(String.join(".", names));
+            return;
+        }
+
+        if (depth > 0) {
+            if (parent.size() < depth) {
+                parent.add(root.name);
+            } else {
+                parent.set(depth - 1, root.name);
+            }
+        }
+
+        for (TrieNode child : root.children.values()) {
+            traverse(child, depth + 1, result, parent);
+        }
+    }
+
+    private static class Trie {
+        private TrieNode root = new TrieNode();
+
+        public Trie(Collection<String> names) {
+            for (String name : names) {
+                insert(name);
+            }
+        }
+
+        public void insert(String name) {
+            if (StringUtils.isBlank(name)) {
+                return;
+            }
+
+            String[] names = name.split("\\.");
+            TrieNode parent = this.root;
+            for (int i = 0; i < names.length; i++) {
+                TrieNode node;
+                if (parent.children.containsKey(names[i])) {
+                    node = parent.children.get(names[i]);
+                } else {
+                    node = new TrieNode(names[i]);
+                    parent.children.put(names[i], node);
+                }
+
+                parent = node;
+            }
+
+            parent.isLeaf = true;
```

Comment:
set a 'parent' node to leaf is a little bit confusing to me. How about rename the `parent` to `currentNode`?